### ISSUE TITLE
Conversion of optional numbers results in 0, and ignores default, when given blank string

### ISSUE THREAD

**Issue by cjsauer** [icebob/fastest-validator#335]
Hi there,

As the title says, using `convert: true` for a `type: 'number'` and `optional: true` causes fastest to sanitize the number as `0`, even if `default: null` is specified. Here is an example:

```javascript
const v = new Validator()

const schema = {
  num: {
      type: 'number',
      convert: true,
      optional: true,
      default: null,
  }
}

const check = v.compile(schema)

const obj = { num: '' }

console.log(check(obj)) // => true
console.log(obj.num) // => 0
```

I would expect the last log above to print `null` as opposed to `0`. Without knowing too much about the code, I suspect [the culprit is here](https://github.com/icebob/fastest-validator/blob/105f908c4598dc7875b8eb69dfdc57797a09e0c4/lib/rules/number.js#L17), in the use of `Number`, which indeed results in `0` when given an empty string.

Maybe this is expected behavior, but it surprised me nonetheless. My html forms generally still send `''` as the form encoded value when an input is cleared out.

**Comment by icebob** [comment]
I think the result is correct, because `optional: true` is triggered by `null` or `undefined` and not empty string. Empty string is a correct value and can be converted to number, so the result is `0`. By the way, here is the generated code which looks fine, it process the value in good order:

```js
// Context.fn[1]
function anonymous(value, field, parent, errors, context, label) {
    var origValue = value;

    if (typeof value !== "number") {
        value = Number(value);
    }

    if (typeof value !== "number" || isNaN(value) || !isFinite(value)) {
        errors.push({
            type: "number",
            message: "The '{field}' field must be a number.",
            field: field,
            actual: origValue,
            label: label
        });
        return value;
    }

    return value;
}

// Context.fn[0]
function anonymous(value, field, parent, errors, context, label) {
    if (typeof value !== "object" || value === null || Array.isArray(value)) {
        errors.push({
            type: "object",
            message: "The '{field}' must be an Object.",
            field: field,
            actual: value,
            label: label
        });
        return value;
    }

    var parentObj = value;
    var parentField = field;

    // Field: num
    field = parentField ? parentField + ".num" : "num";
    value = parentObj.num;
    label = undefined;

    if (value === undefined) {
        // allow undefined
    } else if (value === null) {
        // allow null
    } else {
        parentObj.num = context.fn[1](value, field, parentObj, errors, context, label);
    }

    return parentObj;
}

// Main check function
function anonymous(value, context) {
    var errors = [];
    var field;
    var parent = null;
    var label = null;

    if (value === undefined) {
        errors.push({
            type: "required",
            message: "The '{field}' field is required.",
            field: field,
            actual: value,
            label: label
        });
    } else if (value === null) {
        errors.push({
            type: "required",
            message: "The '{field}' field is required.",
            field: field,
            actual: value,
            label: label
        });
    } else {
        context.fn[0](value, field, null, errors, context, label);
    }

    if (errors.length) {
        return errors.map((err) => {
            if (err.message) {
                err.message = context.utils.replace(err.message, /\{field\}/g, err.label || err.field);
                err.message = context.utils.replace(err.message, /\{expected\}/g, err.expected);
                err.message = context.utils.replace(err.message, /\{actual\}/g, err.actual);
            }
            if (!err.label) delete err.label;
            return err;
        });
    }
    return true;
}
```

**Comment by icebob** [comment]
Btw, I faced the same issue in client side, and I solved it on the client side with a little script:

```js
			fixNumericValue(value, entity, field) {
				if (value === "") entity[field] = null;
			},
```

**Comment by cjsauer** [comment]
Thanks for the reply. That makes sense. Will close.
